File: NeetCode/Intervals/Minimum_Interval_To_Include_Each_Query/minimum__interval__to__include__each__query.py
from typing import Optional, List
from collections import defaultdict


class Solution:
    def minInterval(self, intervals: List[List[int]], queries: List[int]) -> List[int]:
        intervalMap = defaultdict(int)

        for start, end in intervals:
            currLen = end - start + 1
            for i in range(start, end + 1):
                if i in intervalMap:
                    intervalMap[i] = min(currLen, intervalMap[i])
                else:
                    intervalMap[i] = currLen

        res = []
        for q in queries:
            if q in intervalMap:
                res.append(intervalMap[q])
            else:
                res.append(-1)
        return res

        pass


if __name__ == "__main__":
    sol = Solution()
    intervals = [[1, 3], [2, 3], [3, 7], [6, 6]]
    queries = [2, 3, 1, 7, 6, 8]

    print(sol.minInterval(intervals, queries))

    Output = [2, 2, 3, 5, 1, -1]

    print("Running Solution...")

File: NeetCode/Intervals/Minimum_Interval_To_Include_Each_Query/Minimum_Interval_To_Include_Each_Query.cpp
// Minimum_Interval_To_Include_Each_Query.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    vector<int> minInterval(vector<vector<int>> &intervals, vector<int> &queries)
    {
        unordered_map<int, int> intervalMap;

        for (const auto interval : intervals)
        {
            int currLen = interval[1] - interval[0] + 1;
            for (int i = interval[0]; i <= interval[1]; i++)
            {
                if (intervalMap.count(i))
                {
                    intervalMap[i] = min(currLen, intervalMap[i]);
                }
                else
                {
                    intervalMap[i] = currLen;
                }
            }
        }
        vector<int> results;
        for (const auto q : queries)
        {
            if (intervalMap.count(q))
            {
                results.push_back(intervalMap[q]);
            }
            else
            {
                results.push_back(-1);
            }
        }
        return results;
    }
};
